### PR TITLE
schema: Introduce an IPv6 CIDR for the vm network

### PR DESCRIPTION
**Pull request by EdDev**
**What this PR does / why we need it**:

When using the masquerade binding method, an internal IPv6 network is used
for the pod-vm communication.
This change introduces the ability to customize this network address in
scenarios where the default one cannot be used.

**Special notes for your reviewer**:

~~Tests are missing for both CIDR 4 and 6.~~

**Release note**:
```release-note
Add `vmIPv6NetworkCIDR` under `NetworkSource.pod` to support custom IPv6 CIDR for the vm network when using masquerade binding.
```

### CODE REVIEW

File: staging/src/kubevirt.io/client-go/api/v1/schema.go
Comment (rmohr) [inline staging/src/kubevirt.io/client-go/api/v1/schema.go:1090]: It looks like this is not used anywhere. Is there maybe a commit missing?

Comment (EdDev) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go:1090]: Originally I wanted to add a test, see it fails and then link it to the usage and see it passes.
But it seems I need to do some work on the network tests in general and introduce tests for these API/s afterwards.

I added the usage now, not to confuse anyone.

**Comment by AlonaKaplan** [comment]
The change lgtm. Would love to see e2e test. Do you plan to add it in a separate PR?

**Comment by EdDev** [comment]
> The change lgtm. Would love to see e2e test. Do you plan to add it in a separate PR?

Thanks!
I plan to add them here once I arrange the current mess.

**Comment by EdDev** [comment]
#3460 introduces tests for the IPv4 CIDR scenario.
Unfortunately, doing the same for IPv6 is more challenging and requires modifications to the existing code (i.e. refactoring).

**Comment by EdDev** [comment]
> Can you add tests?

I commented about it (see above). The current implementation does not allow doing it without a refactor effort. However, https://github.com/kubevirt/kubevirt/pull/3454 may do the job and then it will be easy.

**Comment by rmohr** [comment]
> I commented about it (see above). The current implementation does not allow doing it without a refactor effort. However, #3454 may do the job and then it will be easy.

Ok.

/hold

Let's wait for tests. Feel free to cancel the hold once they are there.

File: tests/network/vmi_networking.go
Comment (maiqueb) [inline tests/network/vmi_networking.go:748]: IIUC, you're doing this to test the default value. Good.

I don't understand why you would change the value of an input parameter though. I'd have something like

```golang
expectedNetworkCIDR := networkCIDR
if networkCIDR == "" {
    expectedNetworkCIDR = api.DefaultVMIpv6CIDR
}
```
and compare against that new variable. It is more explicit and readable, does not change the value of an input parameter (something which I consider a bad practice), and cost a little bit of memory.

Will not make a fuss over this though.

File: api/openapi-spec/swagger.json
Comment (maiqueb) [inline api/openapi-spec/swagger.json:10769]: The `if not specified` is redundant. `Defaults to` only has meaning when the parameter is not specified.

Not sure what's worse, being redundant or break consistency (the `vmNetworkCIDR` parameter features the same wording).

File: pkg/virt-operator/resource/generate/components/validations_generated.go
Comment (maiqueb) [inline pkg/virt-operator/resource/generate/components/validations_generated.go:2811]: Ditto.

**Review by maiqueb — CHANGES_REQUESTED** [review]
Must remove the focus instruction in the tests; all other comments are opinionated.

File: tests/network/vmi_networking.go
Comment (EdDev) [reply tests/network/vmi_networking.go:735]: Done

File: tests/network/vmi_networking.go
Comment (EdDev) [reply tests/network/vmi_networking.go:748]: It is actually worse, I should not have used it here.
Fixed it now by passing it to where it is really needed. Still I update the input arg to a default, it feels natural to me, it is just a way to implement defaults in Go.

File: api/openapi-spec/swagger.json
Comment (EdDev) [reply api/openapi-spec/swagger.json:10769]: As this is part of the API and potential documentation is generated from it, I prefer to stay safe and use the current format which mimics IPv4. But I do agree with you.
I will try to propose the text change in a separate PR to align both and see how it goes.

Comment (maiqueb) [reply api/openapi-spec/swagger.json:10769]: Fair enough; thanks.

File: tests/network/vmi_networking.go
Comment (maiqueb) [reply tests/network/vmi_networking.go]: I don't think so. Mutating an input argument is, imo, a bad idea, and bad practice. 

But hey, if the language cared about that, they'd have not allowed for you to mutate variables that are not declared w/ var. 

I'd just prefer to have immutable constructs everywhere; as I said, your call, not making a fuss over this.

Comment (EdDev) [reply tests/network/vmi_networking.go]: It is one of the ways to implement a "default" and I have not seen anyone reasoning why this is not allowed.
In C/C++ you have the `const` keyword, in others you can have a default in the signature. Here we can do this.

If you think that it is bad practice, I would love to know the arguments for it. You can take this case as an example if you like.

File: staging/src/kubevirt.io/client-go/api/v1/schema.go
Comment (rmohr) [inline staging/src/kubevirt.io/client-go/api/v1/schema.go:1354]: nit: `json:vmIPv6NetworkCIDR` should be the more correct field name.

Comment (EdDev) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go:1354]: I agree, but this is following the IPv4 format and I am not sure if it is better to break the consistency here.

File: tests/network/vmi_networking.go
Comment (maiqueb) [reply tests/network/vmi_networking.go]: > It is one of the ways to implement a "default" and I have not seen anyone reasoning why this is not allowed.
In C/C++ you have the const keyword, in others you can have a default in the signature. Here we can do this.

I think we're actually saying the same :) 

Above I wrote: "if the language cared about that, they'd have not allowed for you to mutate variables that are not declared w/ var."

> If you think that it is bad practice, I would love to know the arguments for it. You can take this case as an example if you like.

If you never mutate a variable, you will know at all times what it's value is. You can move code around more easily, since no matter where you put it, it'll have the same thing in it. Plenty more arguments about immutable variables / data structure can be found online.

But, again, your call. You want to mutate an input argument, I won't stand in your way. I am just pointing out that you could achieve the same thing in a safer way. Up to you.

Comment (EdDev) [reply tests/network/vmi_networking.go]: I agree that mutating an argument with different values in the same scope is problematic.
In this case, I think it is normalized in the begining and not in the middle, also not used for two different things (like the original code which was wrong).
In my opinion, with your reasoning in mind, the rule should be that a value of an argument should be stable from the first use. Until the first use, it can be adjusted, like with the default pattern.

File: staging/src/kubevirt.io/client-go/api/v1/schema.go
Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go:1354]: I could not find an `Ipv4` field in the schema, could you point me to the field which you mean?

Comment (EdDev) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go:1354]: Ahh... I am wrong here. I thought about something else completely. Sorry.
Changing...

File: tests/network/vmi_networking.go
Comment (maiqueb) [reply tests/network/vmi_networking.go]: > I agree that mutating an argument with different values in the same scope is problematic.
> In this case, I think it is normalized in the begining and not in the middle, also not used for two different things (like the original code which was wrong).
> In my opinion, with your reasoning in mind, the rule should be that a value of an argument should be stable from the first use. Until the first use, it can be adjusted, like with the default pattern.

I keep thinking it should be normalized at creation time, not when it is used. E.g. it's first use may move. And the function behavior will change. 

It's creation time may move, but the compiler will blow up on your face if you use it before creating it, so it is safer.

And IMHO, 'defaulting' things has a time do be done: at creation time, nor when you don't find a value in it. I guess this is another facet of the discussion we're having :) 

But again: your call; if the language was meant to be used as I'm describing, it would have a feature set to enforce it.

File: staging/src/kubevirt.io/client-go/api/v1/schema.go
Comment (EdDev) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go]: Done

Comment (rmohr) [reply staging/src/kubevirt.io/client-go/api/v1/schema.go]: :+1:

File: tests/network/vmi_networking.go
Comment (EdDev) [reply tests/network/vmi_networking.go]: At this moment, I think it is fine as is. Having to name one variable/argument right is a hard, trying to name two that mean (almost) the same thing is even harder.

Comment (maiqueb) [reply tests/network/vmi_networking.go]: Fair enough; thanks.

**Review by maiqueb — APPROVED** [review]
Thanks.

**Pull request merged**
